File: src/oss_pipeline/report.py
"""A single human-readable status page, regenerated on every watch cycle.

The pipeline already writes proposal reports and a ledger, but neither answers
the question actually being asked day to day: what is happening with my open
PRs, and is anything waiting on me? Logs do not answer it either -- they are
append-only and you have to reconstruct the present from them.
"""

from __future__ import annotations

from datetime import datetime, timezone

from . import ghapi, policy, replies, store
from .identity import ROOT, load_identity
from .models import Status

OUT = ROOT / "reports" / "STATUS.md"

PR_QUERY = """
query($owner:String!, $name:String!, $number:Int!) {
  repository(owner:$owner, name:$name) {
    pullRequest(number:$number) {
      state merged isDraft mergeable updatedAt url title
      reviewDecision
      commits(last:1) { nodes { commit { statusCheckRollup {
        state contexts(first:100) { nodes {
          __typename
          ... on CheckRun { name conclusion }
          ... on StatusContext { context state }
        } } } } } }
      reviews(last:10) { nodes { state author { login } submittedAt } }
      comments(last:5) { nodes { author { login } createdAt } }
    }
  }
}
"""


def _pr_state(cand) -> dict:
    owner, name = cand.repo.split("/")
    pr = ghapi.graphql(PR_QUERY, owner=owner, name=name,
                       number=cand.pr_number)["repository"]["pullRequest"]
    nodes = pr["commits"]["nodes"]
    rollup = (nodes[0]["commit"].get("statusCheckRollup") or {}) if nodes else {}
    counts: dict[str, int] = {}
    failing: list[str] = []
    for ctx in (rollup.get("contexts", {}) or {}).get("nodes", []) or []:
        if ctx.get("__typename") == "CheckRun":
            key = (ctx.get("conclusion") or "pending").lower()
            if key in ("failure", "timed_out", "cancelled"):
                failing.append(ctx["name"])
        else:
            key = (ctx.get("state") or "pending").lower()
            if key in ("failure", "error"):
                failing.append(ctx.get("context", "?"))
        counts[key] = counts.get(key, 0) + 1
    return {"pr": pr, "counts": counts, "failing": failing}


def build() -> str:
    ident = load_identity()
    now = datetime.now(timezone.utc)
    L = [f"# Pipeline status", "", f"_generated {now:%Y-%m-%d %H:%M} UTC_", ""]

    live = store.by_status(*policy.OPEN_STATUSES)
    actions: list[str] = []

    L += ["## Open pull requests", ""]
    if not live:
        L += ["_none open_", ""]
    for cand in live:
        try:
            info = _pr_state(cand)
        except Exception as exc:
            L += [f"- **{cand.repo}#{cand.pr_number}** — could not read: {str(exc)[:80]}", ""]
            continue
        pr, counts = info["pr"], info["counts"]
        ci = ", ".join(f"{n} {k}" for k, n in sorted(counts.items())) or "no checks"
        decision = pr.get("reviewDecision") or "no review yet"
        L += [
            f"### [{cand.repo}#{cand.pr_number}]({pr['url']})",
            f"_{pr['title']}_", "",
            f"- state **{pr['state']}**"
            + (" · **MERGED**" if pr["merged"] else "")
            + f" · review: **{decision}** · mergeable: {pr.get('mergeable')}",
            f"- CI: {ci}",
        ]
        if info["failing"]:
            L.append(f"- **FAILING:** {', '.join(info['failing'][:6])}")
            actions.append(f"CI failing on {cand.repo}#{cand.pr_number}: "
                           f"{info['failing'][0]}")
        latest = (pr.get("comments", {}).get("nodes") or [])
        if latest:
            c = latest[-1]
            who = (c.get("author") or {}).get("login", "?")
            L.append(f"- last comment: **{who}** at {c['createdAt'][:16].replace('T', ' ')}")
            if who != ident.login:
                actions.append(f"{who} commented on {cand.repo}#{cand.pr_number}")
        pending = replies.pending(cand)
        if pending:
            drafted = sum(1 for p in pending if p.get("draft"))
            L.append(f"- **{len(pending)} item(s) awaiting your reply** "
                     f"({drafted} drafted) — `pipeline replies`")
            actions.append(f"{len(pending)} reply(ies) queued on {cand.repo}#{cand.pr_number}")
        L.append("")

    # Two different things, kept apart on purpose. A section headed "needs you"
    # that lists machine-queued work teaches you to skip the section, which is
    # exactly when a real request gets missed.
    proposed = store.by_status(Status.PROPOSED)
    approved = store.by_status(Status.APPROVED)
    everything = store.all_candidates()

    needs_you = []
    for c in proposed:
        if c.blockers:
            needs_you.append(
                (f"**{c.repo}#{c.issue}** — {c.blockers[0][:90]}", None))
        else:
            needs_you.append(
                (f"**{c.repo}#{c.issue}** {c.title[:58]} — approve or reject",
                 f"pipeline approve {c.slug}"))
    for c in approved:
        if c.blockers:
            needs_you.append((f"**{c.repo}#{c.issue}** — {c.blockers[0][:90]}", None))

    if needs_you:
        L += ["## Needs you", ""]
        for text, cmd in needs_you:
            L.append(f"- {text}")
            if cmd:
                L.append(f"  `{cmd}`")
        L.append("")

    queued = [c for c in approved if not c.blockers]
    if queued:
        L += ["## Queued — no action needed", ""]
        for c in queued:
            caps = policy.check_caps(everything, c)
            why = caps[0] if caps else "will run on the next scheduled cycle (09:30 daily)"
            L.append(f"- **{c.repo}#{c.issue}** {c.title[:52]}")
            L.append(f"  {why}")
        L.append("")

    if actions:
        L = L[:3] + ["## Needs attention", "",
                     *[f"- {a}" for a in actions], ""] + L[3:]
    else:
        L = L[:3] + ["_Nothing needs your attention._", ""] + L[3:]

    from . import ledger
    d = ledger.compute()
    L += [
        "## Totals", "",
        f"- merged **{d['counts']['merged']}** · closed {d['counts']['closed']} · "
        f"open {d['counts']['open']} · tracked {d['counts']['tracked']}",
        f"- merge rate {d['merge_rate']:.0%}",
        "",
    ]
    return "\n".join(L)


def write() -> str:
    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(build())
    return str(OUT)

File: internal/report/report.go
// Package report renders the single status page, regenerated on every watch
// cycle.
//
// The pipeline already writes proposal reports and a ledger, and neither
// answers the question actually being asked day to day: what is happening with
// my open pull requests, and is anything waiting on me? Logs do not answer it
// either -- they are append-only, and the present has to be reconstructed from
// them.
//
// Rendering is separated from gathering so the layout can be tested without a
// network, and so a repository that fails to read degrades to one line instead
// of taking the page down.
package report

import (
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/vimalyad/osspipeline/internal/ledger"

	"github.com/vimalyad/osspipeline/internal/text"
)

// PRView is one open pull request as the page needs it.
type PRView struct {
	Repo      string
	Number    int
	Title     string
	URL       string
	State     string
	Merged    bool
	Review    string
	Mergeable string

	CheckCounts map[string]int
	Failing     []string

	LastCommentBy string
	LastCommentAt string

	PendingReplies int
	DraftedReplies int

	// Err is set when this pull request could not be read. One unreadable
	// repository must cost one line, not the whole page.
	Err string
}

// Item is something waiting on the user, or something already queued.
type Item struct {
	Text string
	// Cmd is the command that resolves it, when one exists. A blocker has
	// none: it needs a human action outside the pipeline.
	Cmd string
}

// Input is everything the page renders from.
type Input struct {
	Generated time.Time
	// Login is the pipeline's own GitHub account, used to tell the user's own
	// comments from a maintainer's.
	Login    string
	Open     []PRView
	NeedsYou []Item
	Queued   []Item
	Ledger   ledger.Ledger
}

// Render produces the page.
func Render(in Input) string {
	var b strings.Builder
	p := func(f string, a ...any) { fmt.Fprintf(&b, f+"\n", a...) }

	p("# Pipeline status")
	p("")
	p("_generated %s UTC_", in.Generated.UTC().Format("2006-01-02 15:04"))
	p("")

	// Attention first, because it is the reason the page is opened.
	if att := attention(in); len(att) > 0 {
		p("## Needs attention")
		p("")
		for _, a := range att {
			p("- %s", a)
		}
		p("")
	} else {
		p("_Nothing needs your attention._")
		p("")
	}

	p("## Open pull requests")
	p("")
	if len(in.Open) == 0 {
		p("_none open_")
		p("")
	}
	for _, pr := range in.Open {
		if pr.Err != "" {
			p("- **%s#%d** — could not read: %s", pr.Repo, pr.Number, text.Clip(pr.Err, 80))
			p("")
			continue
		}
		p("### [%s#%d](%s)", pr.Repo, pr.Number, pr.URL)
		p("_%s_", pr.Title)
		p("")
		merged := ""
		if pr.Merged {
			merged = " · **MERGED**"
		}
		p("- state **%s**%s · review: **%s** · mergeable: %s",
			pr.State, merged, orElse(pr.Review, "no review yet"), orElse(pr.Mergeable, "unknown"))
		p("- CI: %s", checkSummary(pr.CheckCounts))
		if len(pr.Failing) > 0 {
			p("- **FAILING:** %s", strings.Join(first(pr.Failing, 6), ", "))
		}
		if pr.LastCommentBy != "" {
			p("- last comment: **%s** at %s", pr.LastCommentBy, pr.LastCommentAt)
		}
		if pr.PendingReplies > 0 {
			p("- **%d item(s) awaiting your reply** (%d drafted) — `pipeline replies`",
				pr.PendingReplies, pr.DraftedReplies)
		}
		p("")
	}

	// Two different things, kept apart on purpose. A section headed "needs
	// you" that also lists machine-queued work teaches the reader to skip the
	// section, which is exactly when a real request gets missed.
	if len(in.NeedsYou) > 0 {
		p("## Needs you")
		p("")
		for _, it := range in.NeedsYou {
			p("- %s", it.Text)
			if it.Cmd != "" {
				p("  `%s`", it.Cmd)
			}
		}
		p("")
	}
	if len(in.Queued) > 0 {
		p("## Queued — no action needed")
		p("")
		for _, it := range in.Queued {
			p("- %s", it.Text)
			if it.Cmd != "" {
				p("  %s", it.Cmd)
			}
		}
		p("")
	}

	p("## Totals")
	p("")
	c := in.Ledger.Counts
	p("- merged **%d** · closed %d · open %d · tracked %d", c.Merged, c.Closed, c.Open, c.Tracked)
	p("- merge rate %.0f%%", in.Ledger.MergeRate*100)
	p("")
	if len(in.Ledger.Inconsistent) > 0 {
		p("- %d record(s) disagree with their own history — see the ledger",
			len(in.Ledger.Inconsistent))
		p("")
	}
	return b.String()
}

// attention is the short list at the top: only things a person has to act on.
func attention(in Input) []string {
	var out []string
	for _, pr := range in.Open {
		if pr.Err != "" {
			continue
		}
		if len(pr.Failing) > 0 {
			out = append(out, fmt.Sprintf("CI failing on %s#%d: %s", pr.Repo, pr.Number, pr.Failing[0]))
		}
		// Our own comment is not news. Reporting it trains the reader to
		// ignore the line that matters.
		if pr.LastCommentBy != "" && !strings.EqualFold(pr.LastCommentBy, in.Login) {
			out = append(out, fmt.Sprintf("%s commented on %s#%d", pr.LastCommentBy, pr.Repo, pr.Number))
		}
		if pr.PendingReplies > 0 {
			out = append(out, fmt.Sprintf("%d reply(ies) queued on %s#%d",
				pr.PendingReplies, pr.Repo, pr.Number))
		}
	}
	return out
}

func checkSummary(counts map[string]int) string {
	if len(counts) == 0 {
		return "no checks"
	}
	keys := make([]string, 0, len(counts))
	for k := range counts {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, 0, len(keys))
	for _, k := range keys {
		parts = append(parts, fmt.Sprintf("%d %s", counts[k], k))
	}
	return strings.Join(parts, ", ")
}

func first(ss []string, n int) []string {
	if len(ss) > n {
		return ss[:n]
	}
	return ss
}

func orElse(s, fallback string) string {
	if strings.TrimSpace(s) == "" {
		return fallback
	}
	return s
}
